pkg/controller/master/upgradelog/controller.go

```diff
@@ -0,0 +1,478 @@
+	ctlcorev1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+const (
+	harvesterUpgradeLogLabel            = "harvesterhci.io/upgradeLog"
+	harvesterUpgradeLogStorageClassName = "harvester-longhorn"
+	harvesterUpgradeLogVolumeMode       = corev1.PersistentVolumeFilesystem
+	upgradeLogNamespace                 = "harvester-system"
+
+	upgradeStateLabel                = "harvesterhci.io/upgradeState"
+	UpgradeStateLoggingInfraPrepared = "LoggingInfraPrepared"
+
+	// Annotations for the sub-components of logging infrastructure
+	upgradeLogFluentBitAnnotation     = "harvesterhci.io/fluentBit"
+	upgradeLogFluentdAnnotation       = "harvesterhci.io/fluentd"
+	upgradeLogFluentBitReady          = "FluentBitReady"
+	upgradeLogFluentdReady            = "FluentdReady"
+	upgradeLogClusterFlowAnnotation   = "harvesterhci.io/clusterFlow"
+	upgradeLogClusterOutputAnnotation = "harvesterhci.io/clusterOutput"
+	upgradeLogClusterFlowReady        = "ClusterFlowReady"
+	upgradeLogClusterOutputReady      = "ClusterOutputReady"
+	upgradeLogDownloaderAnnotation    = "harvesterhci.io/downloader"
+	upgradeLogDownloaderReady         = "DownloaderReady"
+)
+
+type handler struct {
+	ctx                 context.Context
+	namespace           string
+	clusterFlowClient   ctlloggingv1.ClusterFlowClient
+	clusterOutputClient ctlloggingv1.ClusterOutputClient
+	daemonSetClient     ctlappsv1.DaemonSetClient
+	daemonSetCache      ctlappsv1.DaemonSetCache
+	deploymentClient    ctlappsv1.DeploymentClient
+	jobClient           ctlbatchv1.JobClient
+	jobCache            ctlbatchv1.JobCache
+	loggingClient       ctlloggingv1.LoggingClient
+	pvcClient           ctlcorev1.PersistentVolumeClaimClient
+	statefulSetClient   ctlappsv1.StatefulSetClient
+	statefulSetCache    ctlappsv1.StatefulSetCache
+	upgradeClient       ctlharvesterv1.UpgradeClient
+	upgradeCache        ctlharvesterv1.UpgradeCache
+	upgradeLogClient    ctlharvesterv1.UpgradeLogClient
+	upgradeLogCache     ctlharvesterv1.UpgradeLogCache
+}
+
+func (h *handler) OnUpgradeLogChange(_ string, upgradeLog *harvesterv1.UpgradeLog) (*harvesterv1.UpgradeLog, error) {
+	if upgradeLog == nil {
+		return upgradeLog, nil
+	}
+
+	if harvesterv1.UpgradeLogReady.GetStatus(upgradeLog) == "" {
+		logrus.Infof("[%s] Initialize upgradeLog %s/%s", upgradeLogControllerName, upgradeLog.Namespace, upgradeLog.Name)
+
+		toUpdate := upgradeLog.DeepCopy()
+		harvesterv1.UpgradeLogReady.CreateUnknownIfNotExists(toUpdate)
+
+		logrus.Infof("[%s] Deploy logging-operator", upgradeLogControllerName)
+		harvesterv1.OperatorDeployed.CreateUnknownIfNotExists(toUpdate)
+
+		// NOTE: As of v1.1.1, the logging-operator is by default deployed (rancher-logging), so we set the condition to true directly.
```

Comment:
In v1.2.0, the logging feature will be `addon`, it may not be enabled by default

Reply:
So basically, along with the conversion of the logging feature, there are two scenarios that we need to concern of:

1. When users upgrade Harvester from v1.1.1 to v1.1.2

Users cannot benefit from the enhancement due to the nature of Harvester's upgrade design, i.e., the newly introduced controllers will be effective only after the apply-manifest phase of an upgrade. Also, the hooking point of UpgradeLog creation is at the beginning of an upgrade, and we will miss that timing. But we still want to prepare this enhancement for the next upgrade (to v1.2.0)

2. When users upgrade Harvester from v1.1.2 to v1.2.0

This is the time things become complicated. I've checked the related PRs(#3380 and harvester/harvester-installer#402), and it seems there will be a deletion of the `rancher-logging` ManagedChart during an upgrade. If we leverage the existing `rancher-logging/-crd` ManagedCharts, logs can only be collected until the deletion. The entire logging infrastructure will be removed due to the deletion of the logging-operator and its CRDs.

---

So our best shot might be "deploy another logging ManagedChart dedicated to the upgrade observability enhancement, remove it after the upgrade ended."

What do you think?

Reply:
that is good idea.

we may utilize managedchart/helm or even addon to deploy, instead of coding the deployment in controller, then more feature changes may reside in the .yaml file

logging components use certain amount of resources, they also need to be added into preflight check when it is enabled

Reply:
> we may utilize managedchart/helm or even addon to deploy, instead of coding the deployment in controller, then more feature changes may reside in the .yaml file

Agreed. But there might be CRDs or other clustered resource conflicts while installing the charts. I just experimented with this: installing logging-operator in a namespace different from `cattle-logging-system` with helm on a v1.1.1 Harvester node. It complained a PodSecurityPolicy already exists, thus aborting the installation.

```
$ helm upgrade --install logging-operator banzaicloud-stable/logging-operator --namespace=harvester-system
Release "logging-operator" does not exist. Installing it now.
W0118 16:17:03.085337    4036 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
Error: rendered manifests contain a resource that already exists. Unable to continue with install: PodSecurityPolicy "psp.logging-operator" in namespace "" exists and cannot be imported into the current release: invalid ownership metadata; annotation validation error: key "meta.helm.sh/release-name" must equal "logging-operator": current value is "rancher-logging"; annotation validation error: key "meta.helm.sh/release-namespace" must equal "harvester-system": current value is "cattle-logging-system"
```

We might encounter similar issues in a v1.2.0 environment and upgrade to a future release, say v1.2.1:

- User initiates the upgrade with the logging addon enabled
- User initiates the upgrade with the logging addon disabled

So, in general, the UpgradeLog controller should have the ability to detect the existence of the logging-operator and its CRDs to decide whether to deploy the logging ManageChart or not.

> logging components use certain amount of resources, they also need to be added into preflight check when it is enabled

Unlike the existing `rancher-logging`, we tailored the upgrade logging infrastructure with as few components as possible. It will not collect events, audit logs, and node-level information. With that said, a preflight check is still good to have.

Reply:
Now that in the operator deploying phase, the controller will detect whether the rancher-logging Addon/ManagedChart is installed or not. If both of them are not enabled/installed, the controller will try to deploy the upgradelog-version ManagedChart. After the upgrade is done, the installed ManagedChart will be removed.

Reply:
@w13915984028 Is there any chance to disable the default Logging resource installed by rancher-logging ManagedChart. The UpgradeLog controller will try to install its own version of Logging resource. Right now, they will both exist; therefore, a total of two sets of fluentd and fluentbit installations but only one of them is used by UpgradeLog.

Reply:
After discussing with @w13915984028, we have to patch the rancher-logging chart to disable the default Logging resource. The default Logging resource will bring up a fluentd StatefulSet and a fluentbit DaemonSet that the UpgradeLog will never use. But without configuring any Flows and Outputs, the system resources they consume are neglectable. So we will leave the current implementation as is due to cost efficiency and the release time schedule.